contracts/Relay.sol

```diff
@@ -0,0 +1,431 @@
+        bytes calldata _currentBest,
+        bytes calldata _newBest,
+        uint256 _limit
+    ) external returns (bool);
+}
+
+
+contract Relay is IRelay {
+    using SafeMath for uint256;
+    using BytesLib for bytes;
+    using BTCUtils for bytes;
+    using ValidateSPV for bytes;
+
+    event Extension(bytes32 indexed _first, bytes32 indexed _last);
+    event Reorg(bytes32 indexed _from, bytes32 indexed _to, bytes32 indexed _gcd);
+
+    bytes32 public relayGenesis;
+    bytes32 public bestKnownDigest;
+    bytes32 public lastReorgCommonAncestor;
+    mapping (bytes32 => bytes32) internal previousBlock;
+    mapping (bytes32 => uint256) internal blockHeight;
+
+
+    /// @notice                   Gives a starting point for the relay
+    /// @dev                      We don't check this AT ALL really. Don't use relays with bad genesis
+    /// @param  _genesisHeader    The starting header
+    /// @param  _height           The starting height
+    constructor(bytes memory _genesisHeader, uint256 _height, bytes32 _periodStart) public {
+        require(_genesisHeader.length == 80, "Stop being dumb");
+        bytes32 _genesisDigest = _genesisHeader.hash256();
+        relayGenesis = _genesisDigest;
+        bestKnownDigest = _genesisDigest;
+        lastReorgCommonAncestor = _genesisDigest;
+        blockHeight[_genesisDigest] = _height;
+        blockHeight[_periodStart] = _height.sub(_height % 2016);
+    }
+
+    /// @notice             Adds headers to storage after validating
+    /// @dev                We check integrity and consistency of the header chain
+    /// @param  _anchor     The header immediately preceeding the new chain
+    /// @param  _headers    A tightly-packed list of new 80-byte Bitcoin headers to record
+    /// @param  _internal   True if called internally from addHeadersWithRetarget, false otherwise
+    /// @return             True if successfully written, error otherwise
+    function _addHeaders(bytes memory _anchor, bytes memory _headers, bool _internal) internal returns (bool) {
+        uint256 _height;
+        bytes memory _header;
+        bytes32 _currentDigest;
+        bytes32 _previousDigest = _anchor.hash256();
+
+        uint256 _target = _headers.slice(0, 80).extractTarget();
+        uint256 _anchorHeight = _findHeight(_previousDigest);  /* NB: errors if unknown */
+
+        require(
+            _internal || _anchor.extractTarget() == _target,
+            "Unexpected retarget on external call");
+        require(_headers.length % 80 == 0, "Header array length must be divisible by 80");
+        require(_headers.length / 80 >= 5, "Must supply at least 5 headers");
+
+        /*
+        NB:
+        1. check that the header has sufficient work
+        2. check that headers are in a coherent chain (no retargets, hash links good)
+        3. Store the block connection
+        4. Store the height
+        */
+        for (uint256 i = 0; i < _headers.length / 80; i = i.add(1)) {
```

Comment:
nit: move `_headers.length / 80` in a variable outside of the loop. The EVM SLOADs array lengths per iteration as it cannot know if the length has changed in the loop. Check this in remix:


```Solidity

pragma solidity ^0.5.1;

contract C {
    uint256[] x;
    
    constructor() public {
        for (uint256 i = 0 ; i < 50; i++ ) {
            x.push(i);
        } 
    }
    
    function cheaper() public {
        uint256 y = 0;
        uint256 len = x.length;
        for (uint256 i = 0 ; i < len; i++ ) {
            y+=1;
        }
    }
    
    function costly() public {
        uint256 y = 0;
        for (uint256 i = 0 ; i < x.length; i++ ) {
            y+=1;
        }
    }
}
```

Reply:
Given that this is a memory array, presumably it `MLOAD`s it?